File: common/src/main/java/juuxel/adorn/fluid/FluidKey.java
package juuxel.adorn.fluid;

import com.mojang.serialization.Codec;
import juuxel.adorn.util.EntryOrTag;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.tags.TagKey;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/// A "key" that identifies a fluid or a group of fluids.
///
/// Could be a single fluid, a tag or a list of the former.
///
/// ## JSON format
///
/// A fluid key is one of:
///
/// - a string; if prefixed with `#`, a tag, otherwise a fluid ID
/// - an array of strings as described above
///
/// Examples: `"minecraft:water"`, `"#c:milk"`, `["minecraft:water", "minecraft:lava"]`
public sealed interface FluidKey permits FluidKeyImpl.Simple, FluidKeyImpl.OfArray {
    Codec<FluidKey> CODEC = FluidKeyImpl.CODEC;

    static FluidKey of(Fluid fluid) {
        return new FluidKeyImpl.Simple(new EntryOrTag.OfEntry<>(fluid));
    }

    static FluidKey of(TagKey<Fluid> tag) {
        return new FluidKeyImpl.Simple(new EntryOrTag.OfTag<>(tag));
    }

    /**
     * Returns the set of all fluids matching this key.
     */
    Set<Fluid> getFluids();

    /**
     * Tests whether the fluid matches this key.
     */
    boolean matches(Fluid fluid);

    /**
     * Writes this key to a packet buffer.
     * @see #load
     */
    default void write(FriendlyByteBuf buf) {
        var fluids = getFluids();
        buf.writeVarInt(fluids.size());

        for (Fluid fluid : fluids) {
            buf.writeVarInt(BuiltInRegistries.FLUID.getId(fluid));
        }
    }

    /**
     * Reads a key from a packet buffer.
     * @see #write
     */
    static FluidKey load(FriendlyByteBuf buf) {
        var size = buf.readVarInt();

        if (size == 1) {
            return new FluidKeyImpl.Simple(new EntryOrTag.OfEntry<>(BuiltInRegistries.FLUID.byId(buf.readVarInt())));
        } else {
            List<FluidKeyImpl.Simple> children = new ArrayList<>(size);
            for (int i = 0; i < size; i++) {
                children.add(new FluidKeyImpl.Simple(new EntryOrTag.OfEntry<>(BuiltInRegistries.FLUID.byId(buf.readVarInt()))));
            }
            return new FluidKeyImpl.OfArray(children);
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/fluid/FluidKey.kt
package juuxel.adorn.fluid

import com.mojang.datafixers.util.Either
import com.mojang.datafixers.util.Pair
import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.Decoder
import com.mojang.serialization.DynamicOps
import com.mojang.serialization.Encoder
import net.minecraft.fluid.Fluid
import net.minecraft.network.PacketByteBuf
import net.minecraft.tag.TagKey
import net.minecraft.util.Identifier
import net.minecraft.util.dynamic.Codecs
import net.minecraft.util.registry.Registry

/**
 * A "key" that identifies a fluid or a group of fluids.
 *
 * Could be a single fluid, a tag or a list of the former.
 *
 * ## JSON format
 *
 * A fluid key is one of:
 *
 * - a string; if prefixed with `#`, a tag, otherwise a fluid ID
 * - an array of strings as described above
 *
 * Examples: `"minecraft:water"`, `"#c:milk"`, `["minecraft:water", "minecraft:lava"]`
 */
sealed class FluidKey {
    /**
     * Returns the set of all fluids matching this key.
     */
    abstract fun getFluids(): Set<Fluid>

    /**
     * Tests whether the [fluid] matches this key.
     */
    abstract fun matches(fluid: Fluid): Boolean

    /**
     * Writes this key to a packet buffer.
     * @see load
     */
    fun write(buf: PacketByteBuf) {
        val fluids = getFluids()
        buf.writeVarInt(fluids.size)

        for (fluid in fluids) {
            buf.writeVarInt(Registry.FLUID.getRawId(fluid))
        }
    }

    companion object {
        private val SIMPLE_ENCODER = object : Encoder<Simple> {
            override fun <T> encode(input: Simple, ops: DynamicOps<T>, prefix: T): DataResult<T> =
                ops.mergeToPrimitive(prefix, ops.createString(input.id))
        }

        private val SIMPLE_DECODER = object : Decoder<Simple> {
            override fun <T> decode(ops: DynamicOps<T>, input: T): DataResult<Pair<Simple, T>> =
                ops.getStringValue(input)
                    .map {
                        if (it.startsWith('#')) {
                            OfTag(TagKey.of(Registry.FLUID_KEY, Identifier(it.substring(1))))
                        } else {
                            OfFluid(Registry.FLUID[Identifier(it)])
                        }
                    }
                    .map { Pair.of(it, ops.empty()) }
        }

        private val SIMPLE_CODEC: Codec<Simple> = Codec.of(SIMPLE_ENCODER, SIMPLE_DECODER, "SimpleFluidKey")

        val CODEC: Codec<FluidKey> = Codecs.xor(
            SIMPLE_CODEC,
            SIMPLE_CODEC.listOf().xmap({ OfArray(it) }, { it.children })
        ).xmap(
            { it.map({ x -> x }, { x -> x }) },
            {
                when (it) {
                    is Simple -> Either.left(it)
                    is OfArray -> Either.right(it)
                }
            }
        )

        /**
         * Reads a key from a packet buffer.
         * @see write
         */
        fun load(buf: PacketByteBuf): FluidKey {
            val size = buf.readVarInt()
            fun readFluid(): Fluid = Registry.FLUID[buf.readVarInt()]

            return if (size == 1) {
                OfFluid(readFluid())
            } else {
                OfArray(
                    (1..size).map {
                        OfFluid(readFluid())
                    }
                )
            }
        }
    }

    private sealed class Simple : FluidKey() {
        abstract val id: String
    }

    private data class OfFluid(val fluid: Fluid) : Simple() {
        override val id: String by lazy { Registry.FLUID.getId(fluid).toString() }
        override fun getFluids() = setOf(fluid)

        override fun matches(fluid: Fluid): Boolean =
            fluid === this.fluid
    }

    private data class OfTag(val tag: TagKey<Fluid>) : Simple() {
        override val id = "#${tag.id}"

        override fun getFluids(): Set<Fluid> =
            Registry.FLUID.getOrCreateEntryList(tag).mapTo(HashSet()) { it.value() }

        override fun matches(fluid: Fluid): Boolean =
            fluid.isIn(tag)
    }

    private data class OfArray(val children: List<Simple>) : FluidKey() {
        override fun getFluids(): Set<Fluid> =
            children.flatMapTo(HashSet()) { it.getFluids() }

        override fun matches(fluid: Fluid): Boolean =
            children.any { it.matches(fluid) }
    }
}
